### PR TITLE
Expand log functionality

### PR DESCRIPTION
**Pull request by flashmob**
Solves #21 (WIP)
The log will go a main log and each server will have its own log file too.
Needs to rename / re-open the log files at any time
Will see if a special logrus hook can be used for that.
Will also see if it can minimise logrus imports to only log.go and put things behind interfaces as much possible

### CODE REVIEW

**Comment by flashmob** [comment]
Hmm.. to fix the tests, cannot use the bytes buffer trick anymore - would need to write log entries to a temp file.

**Comment by flashmob** [comment]
Hi @jordanschalm! The way the logs work is about to go through a big change...

### From a high level


When the server first starts, it will log any startup messages to stderr.
Once it's running, it will detach the log from stderr and attach to what's specified in the config.
The config will have the following settings:

`"log_file"` and `"log_level"` on the main (global) level: 

* `log_file` can be any file path, or the following values: "off", "stdout" or "stderr".

* `log_level` value can be: "debug", "info", "error", "warn", "fatal", "panic"

You can also set a separate `log_file` for each server. If no `log_file` was set for the server, it will default to the main log.

So now you can have a main log, and separate logs for each server.

#### Internal details


There is a new log package that encapsulates logrus as much as possible. Previously, logrus import statements were littered all-over, now there's just one or two logrus imports...

guerrilla package has a new `var mainlog` - this is an extended logrus api which you can use to log to the main log. (So far there is a new WithConn function, but you can add more custom ones if needed) 
Next, the server struct has new `log` and `mainlog` fields which store a pointer to HookLoger struct.

If you ever log anything from the server, please either use the log or mainlog fields in the server struct.

#### Log changing mechanism


When a SIG_HUP is caught, the server will detect any changes to the log config and apply these 
to the next connected client.
The log changing mechanism works the same way as the TLS config reloading.
The new logger is stored in an atomic value, and swapped when the next client connects. 
(So the logs will not change half way during a connection, keeping it consistent)

**Comment by jordanschalm** [comment]
Hey! Thanks for the heads up, this will impact the way I'm handling tracking events. I'll pull these changes down next time I work on the dashboard and get it all figured out.

**Comment by flashmob** [comment]
@jordanschalm - no worries, and if you have any questions then don't hesitate to ask... Next time you are free, could you review the new log.go file and see if it's OK? You can see that it declares a Logger interface, which is used everywhere later - so that it would be easier for other to supply their own Logger. Meanwhile the file re-opening is done using a custom logrus hook.
With the event tracking, are you doing it via hook? If so, do you need to inject your hook into the Logger?

File: log/log.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+	hook := LogrusHook{fname: dest}
+	err := hook.setup(dest)
+	return &hook, err
+}
+
+// Setups sets the hook's writer w and file descriptor w
+// assumes the hook.fd is closed and nil
+func (hook *LogrusHook) setup(dest string) error {
+
+	if dest == "" || dest == "stderr" {
```
Comment (jordanschalm) [inline log/log.go]: could use an enum for these string constants

**Comment by flashmob** [comment]
Thanks! Actually, good call with the lock. Looking at the standard library, they always lock first.

**Pull request merged**